packages/semi-ui/_portal/index.tsx

```diff
@@ -31,46 +64,44 @@ class Portal extends PureComponent<PortalProps, PortalState> {
     static propTypes = {
         children: PropTypes.node,
         prefixCls: PropTypes.string,
-        getPopupContainer: PropTypes.func,
+        style: PropTypes.object,
+        initStyle: PropTypes.object,
         className: PropTypes.string,
+
+        getPopupContainer: PropTypes.func,
         didUpdate: PropTypes.func,
     };
 
-    el: HTMLElement;
+    el: HTMLElement | null;
     context: ContextValue;
+
     constructor(props: PortalProps, context: ContextValue) {
         super(props);
+        /**
+         * Unlike other implementations, we initize `container` and `el` immediately to get children's refs in this tick.
+         * See issue: https://github.com/DouyinFE/semi-design/issues/1703
+         * 
+         * We only append `el` to `container` at `componentDidMount` to fix react18 strict mode error.
+         */
         this.state = {
-            container: this.initContainer(context, true)
+            container: this.getContainer(context)
         };
+        this.el = createEl(Object.assign({}, props.style, props.initStyle));
     }
 
     componentDidMount() {
-        const container = this.initContainer(this.context);
-        if (container!==this.state.container) {
-            this.setState({ container });
+        if (!this.el) {
+            this.el = createEl(Object.assign({}, this.props.style, this.props.initStyle));
         }
-    }
-
-    initContainer = (context: ContextValue, catchError = false) => {
-        try {
-            let container: HTMLElement | undefined = undefined;
-            if (!this.el || !this.state?.container || !Array.from(this.state.container.childNodes).includes(this.el)) {
-                this.el = document.createElement('div');
-                const getContainer = this.props.getPopupContainer || context.getPopupContainer || defaultGetContainer;
-                const portalContainer = getContainer();
-                portalContainer.appendChild(this.el);
-                this.addStyle(this.props.style);
-                this.addClass(this.props.prefixCls, context, this.props.className);
-                container = portalContainer;
-                return container;
-            }
-        } catch (e) {
-            if (!catchError) {
-                throw e;
+        const container = this.getContainer();
+        if (this.el.parentElement !== container) {
+            if (container) {
+                container.appendChild(this.el);
```

Comment:
Indeed, it's a tradeoff, and I'm uncertain about which way to choose.

Reply:
I believe the design is sound. React components can be mounted at any node, even if that node isn't attached to a document.

Reply:
Considering React's substantial emphasis on server components, it's advisable to avoid incorporating side effects within the constructor or render methods. It's conceivable that future versions may discourage or disallow such practices.

Reply:
It's up to you to choose a proper time to make a break change. I totally understand. 🤠